extensions/amp-subscriptions/0.1/platform-store.js

```diff
@@ -366,68 +366,114 @@ export class PlatformStore {
+   * @private
+   */
+  getSupportedFactorWeight_(factorName, platform) {
+    return platform.getSupportedScoreFactor(factorName)
+        .then(factorValue => {
+          if (typeof factorValue !== 'number') {
+            return 0;
+          }
+          return this.scoreConfig_[factorName] *
+            Math.min(1, Math.max(-1, factorValue));
+        });
+  }
+
+  /**
+   * Returns most qualified platform. Qualification of a platform is based on
+   * weight. Every platform starts with weight 0 and evaluated against
    * the following parameters,
-   * - user is subscribed with platform (Gives weight 10)
-   * - supports the current viewer (Gives weight 9)
+   * - base weight
+   * - weight factors the platform supports multiploed by score in the config
    *
    * In the end candidate with max weight is selected. However if candidate's
    * weight is equal to local platform, then local platform is selected.
-   * @return {!./subscription-platform.SubscriptionPlatform}
+   * @return {!Promise<!./subscription-platform.SubscriptionPlatform>}
+   * @param {string=} optionalFactor if present only use this factor for calculation
    * @private
    */
-  selectApplicablePlatform_() {
+  selectApplicablePlatform_(optionalFactor) {
     const localPlatform = this.getLocalPlatform();
-    let localWeight = 0;
-    /** @type {!Array<!Object<!./subscription-platform.SubscriptionPlatform, number>>} */
-    const platformWeights = [];
 
     dev().assert(this.areAllPlatformsResolved_(),
         'All platforms are not resolved yet');
 
-    this.getAvailablePlatforms().forEach(platform => {
-      let weight = 0;
+    /// Subscriber wins immediatly.
+    const availablePlatforms = this.getAvailablePlatforms();
+    while (availablePlatforms.length) {
+      const platform = availablePlatforms.pop();
       const entitlement =
           this.getResolvedEntitlementFor(platform.getServiceId());
-
-      // Subscriber wins immediatly.
       if (entitlement.isSubscriber()) {
-        weight += 100000;
+        return Promise.resolve(platform);
       }
+    }
 
-      // Add the base score
-      weight += platform.getBaseScore();
+    return this.getAllPlatformWeights_(optionalFactor)
+        .then(platformWeights => {
+          platformWeights.sort(function(platform1, platform2) {
+            // Force local platform to win ties
+            if (platform2.weight == platform1.weight &&
+              platform2 == localPlatform) {
+              return 1;
+            }
+            return platform2.weight - platform1.weight;
+          });
+          return platformWeights[0].platform;
+        });
+  }
 
-      // If supports the current viewer, gains weight 9
-      if (platform.supportsCurrentViewer()) {
-        weight += this.scoreConfig_['supportsViewer'];
-      }
+  /**
+   * Calculate and return weights for all platforms
+   * @return {!Promise<!Array<!Object<!./subscription-platform.SubscriptionPlatform, number>>>}
```

Comment:
The type doesn't seem to correspond to what's actually returned. The Object can only be keyed by a string, I believe. Maybe it's `Array<{platform: Platform, weight: number}>`?

Reply:
Yep you're right.